### PR TITLE
Resources export, serialisation & deserialisation

### PR DESCRIPTION
**Pull request by mjs**
This pull request covers the export, serialisation and deserialisation of application resources.

Still to come:
* import of resources back into state
* handling of per unit resource revisions
* export and import of resource binaries

### QA

Visual inspection of `juju dump-model` output for a model with a charm which uses resources. More involved QA will be done once the above items have been handled.

Also tested a standard migration to ensure there's no regressions.

### CODE REVIEW

**Comment by mjs** [comment]
!!try!!

**Comment by mjs** [comment]
!!build!!

File: core/description/resource.go [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+		"path":          schema.String(),
+		"description":   schema.String(),
+		"origin":        schema.String(),
+		"fingerprint":   schema.String(),
+		"size":          schema.Int(),
+		"add-timestamp": schema.Time(),
+		"username":      schema.String(),
+	}
+	defaults := schema.Defaults{
+		"add-timestamp": time.Time{},
```
Comment (howbazaar) [inline core/description/resource.go]: It would be easier to have schema.Omit for the timestamp, and then only set it below if it exists, rather than go through zero.

File: core/description/resource_test.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+	s.sliceName = "resources"
+	s.importFunc = func(m map[string]interface{}) (interface{}, error) {
+		return importResources(m)
+	}
+	s.testFields = func(m map[string]interface{}) {
+		m["resources"] = []interface{}{}
+	}
+}
+
+func minimalResourceMap() map[interface{}]interface{} {
```
Comment (howbazaar) [inline core/description/resource_test.go]: Hmm... this isn't a minimal one though is it?

File: featuretests/package_test.go [hunk 3]
```diff
@@ -14,12 +14,18 @@ import (
...
 	jujucmd "github.com/juju/juju/cmd/juju/commands"
+	"github.com/juju/juju/component/all"
 	coretesting "github.com/juju/juju/testing"
 )
 
 var runFeatureTests = flag.Bool("featuretests", true, "Run long-running feature tests.")
 
 func init() {
+	// Required for anything requiring components (e.g. resources).
+	if err := all.RegisterForServer(); err != nil {
```
Comment (howbazaar) [inline featuretests/package_test.go]: This is really icky.

File: resource/api/client/base_test.go [hunk 4]
```diff
@@ -81,6 +81,7 @@ func newResource(c *gc.C, name, username, data string) (resource.Resource, api.R
 	apiRes := api.Resource{
 		CharmResource: api.CharmResource{
 			Name:        name,
+			Description: name + " description",
```
Comment (howbazaar) [inline resource/api/client/base_test.go]: Why were these needed?

File: state/migration_export.go [hunk 5]
```diff
@@ -482,15 +483,25 @@ func (e *exporter) applications() error {
...
 
+	resourcesSt, err := e.st.Resources()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
 	for _, application := range applications {
 		applicationUnits := e.units[application.Name()]
 		leader := leaders[application.Name()]
+		resources, err := resourcesSt.ListResources(application.Name())
```
Comment (howbazaar) [inline state/migration_export.go]: Bah humbug. The only part of export where we do more than one query per collection.

File: state/upgrade_test.go [hunk 6]
```diff
@@ -5,8 +5,7 @@ package state_test
 
 import (
 	"fmt"
-	"time" // Only used for time types.
-
+	// Only used for time types.
```
Comment (howbazaar) [inline state/upgrade_test.go]: Don't need this comment now.

File: core/description/resource.go [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+}
+
+// ResourceRevision represents a revision of an application resource.
+type ResourceRevision interface {
+	Revision() int
+	Type() string
+	Path() string
+	Description() string
+	Origin() string
+	FingerprintHex() string
```
Comment (mjs) [inline core/description/resource.go]: I did this because on resource.Resource the fingerprint is stored as a raw `[]byte` but here we convert it to a hex representation of that. The name is a reminder that a conversion has been performed.

File: featuretests/package_test.go (hunk 3 above)
Comment (mjs) [inline featuretests/package_test.go]: The joys of the "component architecture".

File: state/migration_export.go (hunk 5 above)
Comment (mjs) [inline state/migration_export.go]: I *might* be able to avoid the extra lookups but it'll mean extending or going around the many levels of resources state API.

File: state/upgrade_test.go (hunk 6 above)
Comment (mjs) [inline state/upgrade_test.go]: Interesting. `goimports` preserves comments when it removes an import.

File: core/description/application_test.go [hunk 8]
```diff
@@ -71,6 +77,7 @@ func minimalApplication(args ...ApplicationArgs) *application {
 	u.SetAgentStatus(minimalStatusArgs())
 	u.SetWorkloadStatus(minimalStatusArgs())
 	u.SetTools(minimalAgentToolsArgs())
+	s.setResources([]*resource{minimalResource()})
```
Comment (mjs) [inline core/description/application_test.go]: b/c this way I can reuse minimalResource :)

File: core/description/application.go [hunk 9]
```diff
@@ -45,6 +45,8 @@ type application struct {
 	// unit count will be assumed by the number of units associated.
 	Units_ units `yaml:"units"`
 
+	Resources_ resources `yaml:"resources,omitempty"`
```
Comment (mjs) [inline core/description/application.go]: Fixed

File: core/description/application_test.go [hunk 10]
```diff
@@ -238,3 +245,12 @@ func (s *ApplicationSerializationSuite) TestLeaderValid(c *gc.C) {
 	err := application.Validate()
 	c.Assert(err, gc.ErrorMatches, `missing unit for leader "ubuntu/1" not valid`)
 }
+
+func (s *ApplicationSerializationSuite) TestResourcesAreValidated(c *gc.C) {
+	application := minimalApplication()
+	application.setResources([]*resource{
```
Comment (mjs) [inline core/description/application_test.go]: Fixed

File: core/description/resource.go [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+	"github.com/juju/errors"
+	"github.com/juju/schema"
+)
+
+// Resource represents an application resource.
+type Resource interface {
+	Name() string
+	Revision() int
+	CharmStoreRevision() int
+	AddRevision(ResourceRevisionArgs)
```
Comment (mjs) [inline core/description/resource.go]: Fixed

File: core/description/resource.go [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+// ResourceRevision represents a revision of an application resource.
+type ResourceRevision interface {
+	Revision() int
+	Type() string
+	Path() string
+	Description() string
+	Origin() string
+	FingerprintHex() string
+	Size() int64
+	AddTimestamp() time.Time
```
Comment (mjs) [inline core/description/resource.go]: Done. I went with `Timestamp`

File: core/description/resource.go [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+	Revision           int
+	CharmStoreRevision int
+}
+
+func newResource(args ResourceArgs) *resource {
+	return &resource{
+		Name_:               args.Name,
+		Revision_:           args.Revision,
+		CharmStoreRevision_: args.CharmStoreRevision,
+		Revisions_:          make([]*resourceRevision, 0),
```
Comment (mjs) [inline core/description/resource.go]: Duh. Fixed.

File: core/description/resource.go (hunk 1 above)
Comment (mjs) [inline core/description/resource.go]: Yep, that's slightly better. I'll make a helper and change all the other places in the package that have time pointers going via the zero value.

File: core/description/resource_test.go (hunk 2 above)
Comment (mjs) [inline core/description/resource_test.go]: Close enough :)  One revision and one charm store revision is fairly small and exercises most of the functionality.

File: state/upgrade_test.go (hunk 6 above)
Comment (mjs) [inline state/upgrade_test.go]: Fixed.

File: core/description/util.go [hunk 14]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package description
+
+import "time"
+
+func fieldToTimePtr(fields map[string]interface{}, name string) *time.Time {
```
Comment (howbazaar) [inline core/description/util.go]: I've generally put these helper functions in serialization.go

**Comment by mjs** [comment]
$$merge$$

**Comment by mjs** [comment]
$$merge$$

**Comment by mjs** [comment]
$$merge$$
